File: frontend/src/Types/Puzzle.ts
import { User } from "./User"

export enum PuzzleCondition {
  "Poor" = "Poor",
  "Fair" = "Fair",
  "Good" = "Good",
  "New" = "New",
}

export interface IPuzzle {
  ID:               number
  name: 					 	string 						
  description:    	string 						
  count:    			 	number 	 						
  puzzleCondition:  PuzzleCondition		
  missingPieces: 	  boolean							
  Owner: 				  	User 						
  owner_id:					number
}

File: backend/model/puzzle.go
package model

import "gorm.io/gorm"

// Puzzle struct
type Puzzle struct {
	gorm.Model
	Name 					 	string 						`gorm:"unique_index;not null" json:"name"`
	Description    	string 						`json:"description"`
	Count    			 	int 	 						`json:"count"`
	PuzzleCondition PuzzleCondition		`json:"puzzleCondition"`
	MissingPieces 	bool							`json:"missingPieces"`
	Owner 					*User 						`gorm:"foreignKey:Owner_id"`
	Owner_id				uint 							`json:"owner_id"`
}

type PuzzleCondition int64

const (
	Poor PuzzleCondition = iota
	Fair        
	Good        
	New        
)

func (s PuzzleCondition) String() string {
	switch s {
	case Poor:
		return "Poor"
	case Fair:
		return "Fair"
	case Good:
		return "Good"
	case New:
		return "New"
	}
	return "unknown"
}